File: rust/main/features/analyze/tests/analyze_int_test.rs
//! Integration tests for disk-cleaner-analyze crate.

use std::fs;
use std::io::Write;
use std::path::Path;
use std::sync::atomic::Ordering;

use disk_cleaner_config::{create_scanner, load_config};

fn write_file(root: &Path, relative: &str, content: &[u8]) {
    let full = root.join(relative);
    if let Some(parent) = full.parent() {
        fs::create_dir_all(parent).unwrap();
    }
    let mut f = fs::File::create(&full).unwrap();
    f.write_all(content).unwrap();
}

fn create_test_config(dir: &Path) -> std::path::PathBuf {
    let config_path = dir.join("profiles.toml");
    fs::write(
        &config_path,
        r#"
[settings]
default_profiles = ["node"]

[profiles.node]
name = "Node.js"
marker = "package-lock.json"
type = "remove"
targets = ["node_modules"]
optional_targets = [".next"]
"#,
    )
    .unwrap();
    config_path
}

/// @covers: analyze::run (measures artifact sizes)
#[test]
fn test_analyze_measures_artifact_sizes() {
    let tmp = tempfile::tempdir().unwrap();
    write_file(tmp.path(), "app/package-lock.json", b"");
    write_file(tmp.path(), "app/node_modules/dep.bin", &vec![0u8; 4096]);

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);
    disk_cleaner_analyze::run(&scanner, &["node".to_string()], &config, false);

    let total = scanner.total_size_bytes.load(Ordering::Relaxed);
    assert!(total >= 4096, "Expected at least 4096 bytes measured, got {total}");
}

/// @covers: analyze::run (does not delete files)
#[test]
fn test_analyze_does_not_modify_filesystem() {
    let tmp = tempfile::tempdir().unwrap();
    write_file(tmp.path(), "app/package-lock.json", b"");
    write_file(tmp.path(), "app/node_modules/dep.bin", &vec![0u8; 1024]);

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);
    disk_cleaner_analyze::run(&scanner, &["node".to_string()], &config, false);

    // Files should still exist
    assert!(tmp.path().join("app/node_modules/dep.bin").exists());
}

/// @covers: analyze::run_disk_usage (generic scan)
#[test]
fn test_analyze_disk_usage_scans_directory() {
    let tmp = tempfile::tempdir().unwrap();
    write_file(tmp.path(), "docs/readme.txt", &vec![0u8; 512]);
    write_file(tmp.path(), "data/big.bin", &vec![0u8; 8192]);

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);

    // Should not panic
    disk_cleaner_analyze::run_disk_usage(&scanner, 2);
}

/// @covers: analyze::run (benchmark mode, no build_command)
#[test]
fn test_analyze_benchmark_empty_build_command_does_not_panic() {
    let tmp = tempfile::tempdir().unwrap();
    write_file(tmp.path(), "app/package-lock.json", b"");
    write_file(tmp.path(), "app/node_modules/dep.bin", &vec![0u8; 1024]);

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);
    // benchmark=true with a profile that has no build_command should skip gracefully
    disk_cleaner_analyze::run(&scanner, &["node".to_string()], &config, true);

    // Should not panic, and files should remain untouched
    assert!(tmp.path().join("app/node_modules/dep.bin").exists());
}

/// @covers: analyze::run (empty directory)
#[test]
fn test_analyze_empty_directory_reports_zero() {
    let tmp = tempfile::tempdir().unwrap();

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);
    disk_cleaner_analyze::run(&scanner, &["node".to_string()], &config, false);

    assert_eq!(scanner.total_size_bytes.load(Ordering::Relaxed), 0);
}

File: java/tests/analyze/AnalyzeIntTest.java
package tests.analyze;

import main.features.config.saf.ConfigFactory;
import main.features.analyze.saf.AnalyzeFactory;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;

/** Integration tests for analyze feature. */
public class AnalyzeIntTest {

    static int passed = 0;
    static int failed = 0;

    public static void main(String[] args) {
        System.out.println("=== Analyze Integration Tests ===");
        System.out.println();

        testAnalyzeMeasuresArtifactSizes();
        testAnalyzeDoesNotModifyFilesystem();
        testAnalyzeDiskUsageScansDirectory();
        testAnalyzeBenchmarkEmptyCommandDoesNotPanic();
        testAnalyzeEmptyDirectoryReportsZero();

        System.out.println();
        System.out.println("Results: " + passed + " passed, " + failed + " failed");
        if (failed > 0) System.exit(1);
    }

    static String createTempDir() { var d = System.getProperty("java.io.tmpdir") + File.separator + "dc-analyze-" + System.nanoTime(); new File(d).mkdirs(); return d; }
    static void writeFile(String root, String rel, String content) { var f = new File(root + File.separator + rel); f.getParentFile().mkdirs(); try { var w = new FileWriter(f); w.write(content); w.close(); } catch (IOException e) { throw new RuntimeException(e); } }
    static void writeBinary(String root, String rel, int size) { var f = new File(root + File.separator + rel); f.getParentFile().mkdirs(); try { var o = new FileOutputStream(f); o.write(new byte[size]); o.close(); } catch (IOException e) { throw new RuntimeException(e); } }
    static void deleteRecursive(File f) { if (f.isDirectory()) { var ch = f.listFiles(); if (ch != null) for (var c : ch) deleteRecursive(c); } f.delete(); }
    static void check(String name, boolean cond) { if (cond) { passed++; System.out.println("  PASS  " + name); } else { failed++; System.out.println("  FAIL  " + name); } }

    static String writeConfig(String dir) {
        writeFile(dir, "profiles.toml", "[settings]\ndefault_profiles = [\"node\"]\n\n[profiles.node]\nname = \"Node.js\"\nmarker = \"package-lock.json\"\ntype = \"remove\"\ntargets = [\"node_modules\"]\noptional_targets = [\".next\"]\n");
        return dir + File.separator + "profiles.toml";
    }

    static void testAnalyzeMeasuresArtifactSizes() {
        var dir = createTempDir();
        try {
            writeBinary(dir, "app/package-lock.json", 0);
            writeBinary(dir, "app/node_modules/dep.bin", 4096);
            var config = ConfigFactory.loadConfig(writeConfig(dir));
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            AnalyzeFactory.run(scanner, new String[]{"node"}, config, false);
            check("test_analyze_measures_artifact_sizes", scanner.totalSizeBytes >= 4096);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testAnalyzeDoesNotModifyFilesystem() {
        var dir = createTempDir();
        try {
            writeBinary(dir, "app/package-lock.json", 0);
            writeBinary(dir, "app/node_modules/dep.bin", 1024);
            var config = ConfigFactory.loadConfig(writeConfig(dir));
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            AnalyzeFactory.run(scanner, new String[]{"node"}, config, false);
            check("test_analyze_does_not_modify_filesystem", new File(dir, "app/node_modules/dep.bin").exists());
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testAnalyzeDiskUsageScansDirectory() {
        var dir = createTempDir();
        try {
            writeBinary(dir, "docs/readme.txt", 512);
            writeBinary(dir, "data/big.bin", 8192);
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            AnalyzeFactory.runDiskUsage(scanner, 2);
            check("test_analyze_disk_usage_scans_directory", true);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testAnalyzeBenchmarkEmptyCommandDoesNotPanic() {
        var dir = createTempDir();
        try {
            writeBinary(dir, "app/package-lock.json", 0);
            writeBinary(dir, "app/node_modules/dep.bin", 1024);
            var config = ConfigFactory.loadConfig(writeConfig(dir));
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            AnalyzeFactory.run(scanner, new String[]{"node"}, config, true);
            check("test_analyze_benchmark_empty_command_does_not_panic",
                    new File(dir, "app/node_modules/dep.bin").exists());
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testAnalyzeEmptyDirectoryReportsZero() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeConfig(dir));
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            AnalyzeFactory.run(scanner, new String[]{"node"}, config, false);
            check("test_analyze_empty_directory_reports_zero", scanner.totalSizeBytes == 0);
        } finally { deleteRecursive(new File(dir)); }
    }
}
